ipaserver/plugins/cert.py

```diff
@@ -858,11 +861,16 @@ def execute(self, csr, all=False, raw=False, chain=False, **kw):
                             "subject alt name type %s is forbidden "
                             "for non-user principals") % "RFC822Name"
                     )
+            elif isinstance(gn, cryptography.x509.general_name.IPAddress):
+                san_has_ipaddrs = True
```

Comment:
I see down below, in `_validate_san_ips`, we iterate the SAN names two more times, once to collect
dns names, and again to collect IP addresses.  I propose, then, that we build two sets during *this* iteration
of the SAN names.  So:
```python
elif isinstance(gn, cryptography.x509.general_name.DNSName):
    san_dnsnames.add(gn.value)  # collect dnsname for IPAddress validation
    ... # existing validation
...
elif isinstance(gn, cryptography.x509.general_name.IPAddress):
    san_ipaddrs.add(gn.value)  # collect ip addresses for IPAddress validation
```

Then later, in the call to `_validate_san_ips` you only need to pass in the sets of collected dnsnames and ipaddresses, so we don't need to traverse the `generalnames` again.
```python
if len(san_ipaddrs) > 0:
    _validate_san_ips(san_dnsnames, san_ipaddrs)
```

Reply:
Sounds reasonable.

My goal with the initial patch was to be as non-intrusive as possible to the existing flow when no IP addresses are present in the CSR.

Reply:
Adding one statement to the DNSName case is an acceptable degree of intrusion :)